util/vole.go

```diff
@@ -0,0 +1,297 @@
+	//  both TCP and QUIC (since the latter was suspected to be a culprit
+	//  here).
+	// FIXME: Check if the GW changes port number and where do we get it.
+
+	ma, err := multiaddr.NewMultiaddr(localhost)
+	if err != nil {
+		panic("failed to decode multi-address")
+	}
+
+	// FIXME: Can we reuse the `bitswapNetwork` in the IPFS core node
+	//  constructor? We would need to overwrite the receive delegate
+	//  (which is controlled by the BitSwap struct) or chain them
+	//  somehow. Right now we are re-connecting to ourselves at the tcp/libp2p
+	//  level creating a new host. (Maybe in terms of testing including this
+	//  layer, even at the bypassed localhost level, is useful.)
+	h, err := libp2p.New(ctx,
+		libp2p.Transport(tcp.NewTCPTransport),
+		libp2p.Transport(quic.NewTransport),
+
+		libp2p.Identity(privKey), // ID: 12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN
+		// FIXME: Could we pass the node's key here? It would be the cleanest
+		//  but maybe BW is not prepared to handle a request from self.
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ai, err := peer.AddrInfoFromP2pAddr(ma)
+	if err != nil {
+		return nil, err
+	}
+
+	hctx, _ := context.WithTimeout(ctx, time.Second*1) // Basic connection timeout for localhost.
+	if err := h.Connect(hctx, *ai); err != nil {
+		return nil, err
+	}
+
+	nilRouter, err := nrouting.ConstructNilRouting(nil, nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	bs := bsnet.NewFromIpfsHost(h, nilRouter)
+
+	bs.SetDelegate(rcv)
+
+	return bs, nil
+}
+
+func createCids(n int) map[cid.Cid]bool {
+	cids := make(map[cid.Cid]bool, 2*n)
+
+	// FIXME: ID hashes seem to be present by default,
+	//   good for testing purposes. (Also they will count
+	//   as `Work` size in the task entry for the priority
+	//   as the size of the CID/block.)
+	//   We should have different sizes of these.
+	v1RawIDPrefix := cid.Prefix{
+		Version: 1,
+		Codec:   cid.Raw,
+		MhType:  mh.IDENTITY,
+	}
+
+	// FIXME: Does raw allow me to create invalid SHA-256 CIDs?
+	//  It seems so: this guarantees nonexistence and helps to test the
+	//  counterpart of the ID CIDs which will always be found.
+	v1RawSha256Prefix := cid.Prefix{
+		Version:  1,
+		Codec:    cid.Raw,
+		MhType:   mh.SHA2_256,
+		MhLength: 20, // To guarantee nonexistence shorten the 32-byte SHA length.
+	}
+
+	for i := 0; i < n; i++ {
+		c, err := v1RawIDPrefix.Sum([]byte(strconv.FormatUint(uint64(i), 10)))
+		if err != nil {
+			panic("error creating raw ID CID")
+		}
+		cids[c] = false
+		c, err = v1RawSha256Prefix.Sum([]byte(strconv.FormatUint(uint64(i), 10)))
+		if err != nil {
+			panic("error creating raw SHA-256 CID")
+		}
+		cids[c] = false
+	}
+	return cids
+}
+
+func createMsg(requestedCids map[cid.Cid]bool) bsmsg.BitSwapMessage {
+	msg := bsmsg.New(true)
+	// Creating a new full list to replace the previous one since each test
+	// is independent.
+	// FIXME: Check if the above make sense.
```

Comment:
That works (or at least it should). It _won't_ test weird cancel races, but we can look into that if we can't find anything this way.